### PR TITLE
Circle starks

### PR DESCRIPTION
**Pull request by ColoCarletti**
# Circle Starks

This PR introduces the foundational components necessary for the Circle Starks protocol implementation.

## Key Features:

- Mersenne-31 field: Incorporates the Mersenne-31 field, its quadratic and quartic extensions, along with all required operations.
- Circle group: Defines the circle group and implements operations between its elements.
- Sets and cosets: Generates sets and cosets to construct the necessary domains within the circle group.
- CFFT: Implements the Circle Fast Fourier Transform (CFFT) for both interpolation and evaluation of the bivariate polynomial used in the Circle Starks protocol.

## Type of change

- [x] New feature

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [x] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [ ] This change is an Optimization
  - [ ] Benchmarks added/run

### CODE REVIEW

File: math/src/circle/point.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    pub fn antipode(self) -> Self {
+        Self {
+            x: -self.x,
+            y: -self.y,
+        }
+    }
+
+    pub fn eq(a: Self, b: Self) -> bool {
+        a.x == b.x && a.y == b.y
+    }
```
[the remark this answers is not in the document]
Comment (ilitteri) [reply math/src/circle/point.rs:137]: Just found `PartialEq` implementation. We can move this implementation inside `PartialEq`'s and remove this function (it comes for free with the trait as well as `ne`).

File: math/src/circle/errors.rs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#[derive(Debug)]
+pub enum CircleError {
+    InvalidValue,
```
[the remark this answers is not in the document]
Comment (ilitteri) [reply math/src/circle/errors.rs:3]: I just found out that we are using `thiserror` in this repo. It'd be good to use it here.

**Review by ilitteri — COMMENTED** [review]
The PR looks good, left some nits, comments, and suggestions. In general, I'd try to remove every change for the code to panic (even if we know for certain that indexing in some index will always be valid).

File: math/src/circle/point.rs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            Ok(CirclePoint { x, y })
+        } else {
+            Err(CircleError::InvalidValue)
+        }
+    }
+
+    /// Neutral element of the Circle group (with additive notation).
+    pub fn zero() -> Self {
+        Self::new(FieldElement::one(), FieldElement::zero()).unwrap()
+    }
```
[the remark this answers is not in the document]
Comment (diegokingston) [reply math/src/circle/point.rs:74]: This should be a const

[the remark this answers is not in the document]
Comment (ilitteri) [reply math/src/circle/point.rs:74]: Good catch! Just updated my suggestion

[the remark this answers is not in the document]
Comment (nicole-graus) [reply math/src/circle/point.rs:63]: `FieldElement::one()` and `FieldElement::zero()` aren't `const`. I understand they should be, but to fix it we have to change it in every field of Lambdaworks. Maybe we can do it in another PR. If we wanted to hardcode the values, we should add them as parameters for every field we wanted to implement the Circle, I don't know if that is what you are suggesting.

File: math/src/circle/point.rs
[the remark this answers is not in the document]
Comment (ilitteri) [reply math/src/circle/point.rs:1]: Yes!

File: math/src/circle/polynomial.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+    // The cfft returns the evaluations in a certain order, so we permute them to get the natural order.
+    order_cfft_result_naive(&mut coeff)
+}
+
+/// Interpolates the 2^n evaluations of a two-variables polynomial of degree 2^n - 1 on the points of the standard coset of size 2^n.
+/// As a result we obtain the coefficients of the polynomial in the basis: {1, y, x, xy, 2xˆ2 -1, 2xˆ2y-y, 2xˆ3-x, 2xˆ3y-xy,...}
+/// Note that eval has to be a vector of length a power of two 2^n.
+#[cfg(feature = "alloc")]
+pub fn interpolate_cfft(
```
Comment (diegokingston) [inline math/src/circle/polynomial.rs:39]: As an optimization, we could just send the twiddles as input to the function. This is useful when we have to interpolate multiple columns

File: math/src/circle/polynomial.rs [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    fft::cpu::bit_reversing::in_place_bit_reverse_permute,
+    field::{element::FieldElement, fields::mersenne31::field::Mersenne31Field},
+};
+use alloc::vec::Vec;
+
+/// Given the 2^n coefficients of a two-variables polynomial of degree 2^n - 1 in the basis {1, y, x, xy, 2xˆ2 -1, 2xˆ2y-y, 2xˆ3-x, 2xˆ3y-xy,...}
+/// returns the evaluation of the polynomial on the points of the standard coset of size 2^n.
+/// Note that coeff has to be a vector with length a power of two 2^n.
+#[cfg(feature = "alloc")]
+pub fn evaluate_cfft(
```
Comment (diegokingston) [inline math/src/circle/polynomial.rs:18]: Similar comment to the one after this one. Twiddles could be given as inputs

File: math/src/circle/cfft.rs [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+/// The cfft doesn't return the evaluations in the natural order.
+/// For example, if we apply the cfft to 8 coefficients of a polynomial of degree 7 we'll get the evaluations in this order:
+/// [P(x0, y0), P(x2, y2), P(x4, y4), P(x6, y6), P(x7, y7), P(x5, y5), P(x3, y3), P(x1, y1)],
+/// where the even indices are found first in ascending order and then the odd indices in descending order.
+/// This function permutes the slice [0, 2, 4, 6, 7, 5, 3, 1] into [0, 1, 2, 3, 4, 5, 6, 7].
+/// TODO: This can be optimized by performing in-place value swapping (WIP).  
+pub fn order_cfft_result_naive(
+    input: &mut [FieldElement<Mersenne31Field>],
+) -> Vec<FieldElement<Mersenne31Field>> {
+    let mut result = Vec::new();
```
Comment (diegokingston) [inline math/src/circle/cfft.rs:83]: maybe we can use interleave

**Review by nicole-graus — APPROVED** [review]

**Review by ilitteri — COMMENTED** [review]
The PR is looking good! Left more comments.

As a side note, if we use an `unwrap` and we're 100% sure that it won't panic, let's add an explanation either as function docs (this is done under a `/// # Panics` section) or as a comment above the line.

File: math/src/circle/polynomial.rs [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+    icfft(&mut eval_ordered, twiddles);
+
+    // The icfft returns the polynomial coefficients in bit reverse order. So we premute it to get the natural order.
+    in_place_bit_reverse_permute::<FieldElement<Mersenne31Field>>(&mut eval_ordered);
+
+    // The icfft returns all the coefficients multiplied by 2^n, the length of the evaluations.
+    // So we multiply every element that outputs the icfft byt the inverse of 2^n to get the actual coefficients.
+    let factor = (FieldElement::<Mersenne31Field>::from(eval.len() as u64))
+        .inv()
+        .unwrap();
```
[the remark this answers is not in the document]
Comment (nicole-graus) [reply math/src/circle/polynomial.rs:64]: Done.

File: math/src/circle/point.rs [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+#[derive(Debug, Clone)]
+pub struct CirclePoint<F: IsField> {
+    pub x: FieldElement<F>,
+    pub y: FieldElement<F>,
+}
+
+#[derive(Debug)]
+pub enum CircleError {
+    PointDoesntSatisfyCircleEquation,
+}
```
[the remark this answers is not in the document]
Comment (ilitteri) [reply math/src/circle/point.rs:22]: Is it possible that it was not removed from this file?

File: math/src/circle/point.rs [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+            if scalar & 1 == 1 {
+                res = &res + &cur;
+            }
+            cur = cur.double();
+            scalar >>= 1;
+        }
+    }
+}
+
+impl<F: IsField + HasCircleParams<F>> CirclePoint<F> {
```
[the remark this answers is not in the document]
Comment (ilitteri) [reply math/src/circle/point.rs:154]: Is it possible that this impl is still at the bottom?

File: math/src/circle/point.rs [hunk 10]
```diff
@@ -1,82 +1,24 @@
...
-/// (a, b) + (c, d) = (a * c - b * d, a * d + b * c)
+/// (a, b) + (c, d) = (a * c - b * d, a * d + b * c).
+
 #[derive(Debug, Clone)]
 pub struct CirclePoint<F: IsField> {
     pub x: FieldElement<F>,
     pub y: FieldElement<F>,
 }
 
-impl<F: IsField + HasCircleParams<F>> CirclePoint<F> {
```
Comment (ilitteri) [inline math/src/circle/point.rs]: Careful here, I think you've undone part of [e0fa390](https://github.com/lambdaclass/lambdaworks/pull/924/commits/e0fa3904aeb98a3146f4a5f73f05c4f801f76287)'s work

Comment (nicole-graus) [reply math/src/circle/point.rs]: You're right, thank you!

File: math/src/circle/point.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (nicole-graus) [reply math/src/circle/point.rs]: Yes, I fixed it

File: math/src/circle/point.rs [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+impl<F: IsField + HasCircleParams<F>> Add<&CirclePoint<F>> for CirclePoint<F> {
+    type Output = CirclePoint<F>;
+    fn add(self, rhs: &CirclePoint<F>) -> Self::Output {
+        &self + rhs
+    }
+}
+
+/// Multiplication between a point and a scalar (i.e. group operation repeatedly):
+/// (x, y) * n = (x ,y) + ... + (x, y) n-times.
+impl<F: IsField + HasCircleParams<F>> Mul<u128> for CirclePoint<F> {
```
[the remark this answers is not in the document]
Comment (nicole-graus) [reply math/src/circle/point.rs]: Done!

File: math/src/circle/point.rs [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+    fn add(self, rhs: CirclePoint<F>) -> Self::Output {
+        self + &rhs
+    }
+}
+impl<F: IsField + HasCircleParams<F>> Add<&CirclePoint<F>> for CirclePoint<F> {
+    type Output = CirclePoint<F>;
+    fn add(self, rhs: &CirclePoint<F>) -> Self::Output {
+        &self + rhs
+    }
+}
```
[the remark this answers is not in the document]
Comment (nicole-graus) [reply math/src/circle/point.rs]: Done!

**Review by ilitteri — APPROVED** [review]
LGTM. Great work!

**Pull request merged**